src/dispatch/batch.rs

```diff
@@ -0,0 +1,453 @@
+use crate::{
+    dispatch::Dispatcher, world::ResourceId, Accessor, AccessorCow, DynamicSystemData, RunningTime,
+    System, SystemData, World,
+};
+
+/// The `BatchAccessor` is used to notify the main dispatcher of the read and
+/// write resources of the `System`s contained by the `Batch` ("sub systems").
+#[derive(Debug)]
+pub struct BatchAccessor {
+    reads: Vec<ResourceId>,
+    writes: Vec<ResourceId>,
+}
+
+impl BatchAccessor {
+    /// Create the `BatchAccessor`
+    pub fn new(reads: Vec<ResourceId>, writes: Vec<ResourceId>) -> Self {
+        BatchAccessor { reads, writes }
+    }
+}
+
+impl Accessor for BatchAccessor {
+    fn try_new() -> Option<Self> {
+        None
+    }
+
+    fn reads(&self) -> Vec<ResourceId> {
+        self.reads.clone()
+    }
+
+    fn writes(&self) -> Vec<ResourceId> {
+        self.writes.clone()
+    }
+}
+
+/// The `BatchUncheckedWorld` wrap an instance of the world.
+/// You have to specify this as `SystemData` of the `BatchControllerSystem`.
+pub struct BatchUncheckedWorld<'a>(pub &'a World);
+
+impl<'a> DynamicSystemData<'a> for BatchUncheckedWorld<'a> {
+    type Accessor = BatchAccessor;
+
+    fn setup(_accessor: &Self::Accessor, _world: &mut World) {}
+
+    fn fetch(_access: &Self::Accessor, world: &'a World) -> Self {
+        BatchUncheckedWorld(world)
+    }
+}
+
+/// The `BatchController` is the additional trait that a normal System must
+/// implement in order to be used as `BatchControllerSystem`.
+///
+/// Note that the `System` must also implement `Send` because the `Dispatcher`
+/// is by default un-send.
+/// Is safe to implement `Send` and `Sync` because the `BatchAccessor` keep
+/// tracks of all used resources and thus the `System` can be safely executed in
+/// multi thread.
+pub trait BatchController<'a, 'b> {
+    /// Here you must set all the `Resource` types that you want to use directly
+    /// inside the `BatchControllerSystem`.
+    ///
+    /// These have to be specified here, instead of `SystemData` (as
+    /// a normal `System` does) because there is a direct dependecy with the sub
+    /// `System`s `Resource`.
+    ///
+    /// Since the sub `System`s will use the same `Resource' of the
+    /// `BatchControllerSystem` is necessary to drop the references to the
+    /// fetched `Resource` before to dispatch the inner `Dispatcher`.
+    ///
+    /// Now is easy to understand that specify the `BatchControllerSystem`
+    /// `Resource` in the `SystemData` doesn't allow to drop the reference
+    /// before the sub dispatching resulting in `Panic`.
```

Comment:
I would write something like:
```markdown
Because `System::SystemData` is fetched for the whole duration of `System::run` dropping the resources isn't possible and dispatching inner `System`s will result to panic.
```
Except you can actually use `drop` function to drop the resources early... So do we actually need this separation? Or am I missing something here again.

Also you have been using the `BatchControllerSystem` term which doesn't exist in the code (which is why these suggestions), but if we still need the trait should we do `trait BatchDispatcherSystem: System` to make it a thing and force implementor to implement both which is what you need to any way do if you want to make custom batch dispatcher.

Reply:
You could need to fetch a resource each dispatching.

Using a `Drop` just before the dispatch, make the resource lost.
Then in the second iteration you have to fetch the resource from the world (which is not obvious).

The problem in this scenario is that the code is not uniform, and you have to make sure to fetch  only from the second iteration.

Instead I think that is more clear, uniform and flexible fetch it explicitly.

However I've updated the comment that should be more clear now.